x/cash/dynamicfee.go

```diff
@@ -0,0 +1,219 @@
+	var res weave.CheckResult
+	fee, payer, cache, err := d.prepare(ctx, store, tx)
+	if err != nil {
+		return res, err
+	}
+
+	minFee := gconf.Coin(store, GconfMinimalFee)
+
+	// do subtransaction in a function for easier error handling
+	res, err = func() (weave.CheckResult, error) {
+		// shadow with local variables...
+		err := d.chargeFee(cache, payer, fee)
+		if err != nil {
+			return weave.CheckResult{}, errors.Wrap(err, "cannot charge fee")
+		}
+		res, err := next.Check(ctx, cache, tx)
+		// TODO: check RequiredFee here and return an error if insufficient
+		return res, err
+	}()
+
+	// on error, rollback, then take the minfee from the store
+	if err != nil {
+		cache.Discard()
+		// if this fails, we aborted early above, we can just ignore return value
+		// this is 2 ops, not 1, for errors, but done to optimize the success case to use 1 not 2
+		_ = d.chargeFee(store, payer, minFee)
+		// return error from the transaction, not the possible error from minFee deduction
+		return res, err
+	}
+
+	// if success, we commit and update the importance
+	cache.Write()
+	res.GasPayment += toPayment(fee)
+	return res, err
+}
+
+// Deliver verifies and deducts fees before calling down the stack
+func (d DynamicFeeDecorator) Deliver(ctx weave.Context, store weave.KVStore, tx weave.Tx,
+	next weave.Deliverer) (weave.DeliverResult, error) {
+
+	var res weave.DeliverResult
+	fee, payer, cache, err := d.prepare(ctx, store, tx)
+	if err != nil {
+		return res, err
+	}
+
+	minFee := gconf.Coin(store, GconfMinimalFee)
+
+	// do subtransaction in a function for easier error handling
+	res, err = func() (weave.DeliverResult, error) {
+		// shadow with local variables...
+		err := d.chargeFee(cache, payer, fee)
+		if err != nil {
+			return weave.DeliverResult{}, errors.Wrap(err, "cannot charge fee")
+		}
+		res, err := next.Deliver(ctx, cache, tx)
+		// TODO: check RequiredFee here and return an error if insufficient
+		return res, err
+	}()
+
+	// on error, rollback, then take the minfee from the store
+	if err != nil {
+		cache.Discard()
+		// if this fails, we aborted early above, we can just ignore return value
+		// this is 2 ops, not 1, for errors, but done to optimize the success case to use 1 not 2
+		_ = d.chargeFee(store, payer, minFee)
+		// return error from the transaction, not the possible error from minFee deduction
+		return res, err
+	}
+
+	// if success, we commit and update the importance
+	cache.Write()
+	return res, err
+}
+
+func (d DynamicFeeDecorator) chargeFee(store weave.KVStore, src weave.Address, amount x.Coin) error {
+	if amount.IsZero() {
+		return nil
+	}
+	dest := gconf.Address(store, GconfCollectorAddress)
+	return d.ctrl.MoveCoins(store, src, dest, amount)
+}
+
+// prepare is all shared setup between Check and Deliver, one more level above extractFee
+func (d DynamicFeeDecorator) prepare(ctx weave.Context, store weave.KVStore, tx weave.Tx) (fee x.Coin, payer weave.Address, cache weave.KVCacheWrap, err error) {
```

Comment:
pp: named return params should be avoided as a best practice. See https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters